external/opcua_daq_types/CMakeLists.txt

```diff
@@ -0,0 +1,249 @@
+  IMPORT_BSD    "UA_TYPES_DAQDEVICE#${COMPANION_SPECIFICATIONS_DIRPREFIX}/Opc.Ua.Daq.Device.Types.bsd"
+  FILE_CSV      "${COMPANION_SPECIFICATIONS_DIRPREFIX}/Opc.Ua.Daq.Esp.NodeIds.csv"
+  FILE_BSD      "${COMPANION_SPECIFICATIONS_DIRPREFIX}/Opc.Ua.Daq.Esp.Types.bsd"
+  OUTPUT_DIR    "${GENERATE_OUTPUT_DIR}"
+  NAMESPACE_MAP "${NAME_SPACE_INDEX}:https://docs.opendaq.io/specifications/opc-ua/daq/esp"
+  FILE_NS       "${COMPANION_SPECIFICATIONS_DIRPREFIX}/Opc.Ua.Daq.Esp.NodeSet2.xml"
+  DEPENDS       "daqdevice"
+  INTERNAL
+)
+list(APPEND OPC_UA_NODESET_DEPENDENCIES ${PROJECT_NAME}-ns-daqesp)
+list(APPEND OPC_UA_NODESET_SOURCES ${UA_NODESET_DAQESP_SOURCES} ${UA_TYPES_DAQESP_SOURCES})
+list(APPEND OPC_UA_NODESET_HEADERS ${UA_NODESET_DAQESP_HEADERS} ${UA_TYPES_DAQESP_HEADERS} ${GENERATE_OUTPUT_DIR}/daqesp_nodeids.h)
+list(APPEND OPC_UA_NODESET_COMPILE_FLAGS NAMESPACE_DAQESP=${NAME_SPACE_INDEX})
+MATH(EXPR NAME_SPACE_INDEX "${NAME_SPACE_INDEX}+1")
+
+
+set(LOOP_RUN_INDEX 0)
+foreach(VENDOR_SPEC ${OPC_UA_VENDOR_LIST})
+  message(STATUS "Generating DAQ ${VENDOR_SPEC} nodeset...")
+  string(TOLOWER ${VENDOR_SPEC} specName)
+  string(TOUPPER ${VENDOR_SPEC} specNameUpper)
+  list(GET VENDOR_COMPANION_SPECIFICATIONS_DIRPREFIX 0 VENDOR_COMPANION_SPECIFICATION_DIRPREFIX)
+
+  ua_generate_nodeset_and_datatypes(
+    NAME          "daq${specName}"
+    TARGET_PREFIX "${PROJECT_NAME}"
+    IMPORT_BSD    "UA_TYPES#${open62541_NODESET_DIR}/Schema/Opc.Ua.Types.bsd"
+    IMPORT_BSD    "UA_TYPES_DAQBT#${COMPANION_SPECIFICATIONS_DIRPREFIX}/Opc.Ua.Daq.Bt.Types.bsd"
+    IMPORT_BSD    "UA_TYPES_DAQBSP#${COMPANION_SPECIFICATIONS_DIRPREFIX}/Opc.Ua.Daq.Bsp.Types.bsd"
+    IMPORT_BSD    "UA_TYPES_DAQDEVICE#${COMPANION_SPECIFICATIONS_DIRPREFIX}/Opc.Ua.Daq.Device.Types.bsd"
+    IMPORT_BSD    "UA_TYPES_DAQESP#${COMPANION_SPECIFICATIONS_DIRPREFIX}/Opc.Ua.Daq.Esp.Types.bsd"
+    FILE_CSV      "${VENDOR_COMPANION_SPECIFICATION_DIRPREFIX}/Opc.Ua.Daq.Vendor.${VENDOR_SPEC}.NodeIds.csv"
+    FILE_BSD      "${VENDOR_COMPANION_SPECIFICATION_DIRPREFIX}/Opc.Ua.Daq.Vendor.${VENDOR_SPEC}.Types.bsd"
+    OUTPUT_DIR    "${GENERATE_OUTPUT_DIR}"
+    NAMESPACE_MAP "${NAME_SPACE_INDEX}:https://docs.opendaq.io/specifications/opc-ua/daq/vendor/${specName}"
+    FILE_NS       "${VENDOR_COMPANION_SPECIFICATION_DIRPREFIX}/Opc.Ua.Daq.Vendor.${VENDOR_SPEC}.NodeSet2.xml"
+    DEPENDS       "daqesp"
+    INTERNAL
+  )
+  list(APPEND OPC_UA_NODESET_DEPENDENCIES ${PROJECT_NAME}-ns-daq${specName})
+  list(APPEND OPC_UA_NODESET_SOURCES ${UA_NODESET_DAQ${specNameUpper}_SOURCES} ${UA_TYPES_DAQ${specNameUpper}_SOURCES})
+  list(APPEND OPC_UA_NODESET_HEADERS ${UA_NODESET_DAQ${specNameUpper}_HEADERS} ${UA_TYPES_DAQ${specNameUpper}_HEADERS} ${GENERATE_OUTPUT_DIR}/daq${specName}_nodeids.h)
+  list(APPEND OPC_UA_NODESET_COMPILE_FLAGS NAMESPACE_DAQ${specNameUpper}=${NAME_SPACE_INDEX})
+  MATH(EXPR NAME_SPACE_INDEX "${NAME_SPACE_INDEX}+1")
+  MATH(EXPR LOOP_RUN_INDEX "${LOOP_RUN_INDEX}+1")
+endforeach()
```

Comment:
The dependencies and imports are specific to the HBK nodeset, and cannot be custom defined. This template makes the assumption that every nodeset imports all of our nodesets and depends on "daqesp". Also all the import targets are named "IMPORT_BSD" which seems wrong to me.

Reply:
This approach is also just weird to me, honestly. If anything, there should be a folder called "vendor" next to the CMakeLists.txt that contains the code for pulling and importing vendor-specific types. All of this code should be in the CMakeLists.txt document, so that any vendor can import it as they deem is correct - we just need to provide the instructions on which variables need to be filled with what. The file that was modded so extensively by this PR should honestly only have a single line of "add_subdirectory" for the HBK spec if it's added to the repo, and 1 extra "add_subdircectory" if `-DOPENDAQ_USE_VENDOR_OPCUA_NODESET` is enabled.

Beyond that, the loop I added to automatically generate the endpoint registration calls should be there as well. Everything else should be handled in individual CMakeLists.txt by the vendor imo.

Reply:
The Import BSD is a key word delivered by open62451. This is correct. And yes, if you want to use opendaq you need to build up the spec based on our definitions including esp. This is how our connection module is implemented. It cannot just parse other types etc. Every property needs to use our variable definitions. How you are augmenting here results in the fact, that we need to rewrite a lot of the connection module.

Reply:
I need to read before the second reply. I can change it to the way, as you wrote it in your second reply. However, I will need to check first how we solve the problem with namespace index handling.